File: src/test/scala/com/github/tetrisanalyzer/piecemove/ValidPieceMovesForEmptyBoardTest.scala
package com.github.tetrisanalyzer.piecemove

import org.junit.Test
import com.github.tetrisanalyzer.board.Board
import com.github.tetrisanalyzer.piece.PieceS
import com.github.tetrisanalyzer.settings.DefaultGameSettings

import scala.collection.mutable.Set
import scala.collection.mutable.LinkedHashSet
import com.github.tetrisanalyzer.BaseTest
import com.github.tetrisanalyzer.move.Move

class ValidPieceMovesForEmptyBoardTest extends BaseTest {

  @Test def startMove {
    val board = Board(5,5)
    val piece = new PieceS
    val settings = new DefaultGameSettings
    val validPieceMovesForEmptyBoard = new ValidPieceMovesForEmptyBoard(board, piece, settings)

    var pieceMoves = new LinkedHashSet[PieceMove]
    addPieceMoves(validPieceMovesForEmptyBoard.startMove, pieceMoves)

    pieceMoves should be (LinkedHashSet(
      PieceMove(board,piece,Move(0,1, 0)),
      PieceMove(board,piece,Move(0,0, 0)),
      PieceMove(board,piece,Move(0,0, 1)),
      PieceMove(board,piece,Move(0,0, 2)),
      PieceMove(board,piece,Move(0,0, 3)),
      PieceMove(board,piece,Move(1,1, 0)),
      PieceMove(board,piece,Move(1,0, 0)),
      PieceMove(board,piece,Move(1,0, 1)),
      PieceMove(board,piece,Move(1,0, 2)),
      PieceMove(board,piece,Move(1,2, 0)),
      PieceMove(board,piece,Move(1,3, 0)),
      PieceMove(board,piece,Move(1,3, 1)),
      PieceMove(board,piece,Move(1,3, 2)),
      PieceMove(board,piece,Move(1,2, 1)),
      PieceMove(board,piece,Move(1,2, 2)),
      PieceMove(board,piece,Move(1,1, 1)),
      PieceMove(board,piece,Move(1,1, 2)),
      PieceMove(board,piece,Move(0,2, 0)),
      PieceMove(board,piece,Move(0,2, 1)),
      PieceMove(board,piece,Move(0,2, 2)),
      PieceMove(board,piece,Move(0,2, 3)),
      PieceMove(board,piece,Move(0,1, 1)),
      PieceMove(board,piece,Move(0,1, 2)),
      PieceMove(board,piece,Move(0,1, 3))
    ))
  }

  private def addPieceMoves(pieceMove: PieceMove, pieceMoves: Set[PieceMove]) {
    if (pieceMove != null && !pieceMoves.contains(pieceMove)) {
      pieceMoves += pieceMove
      pieceMove.asideAndRotate.foreach(move => addPieceMoves(move, pieceMoves))
      addPieceMoves(pieceMove.down, pieceMoves)
    }
  }
}

File: langs/java/src/test/java/com/github/tetrisanalyzer/piecemove/ValidPieceMovesForEmptyBoardTest.java
package com.github.tetrisanalyzer.piecemove;

import com.github.tetrisanalyzer.board.Board;
import com.github.tetrisanalyzer.move.Move;
import com.github.tetrisanalyzer.piece.Piece;
import com.github.tetrisanalyzer.settings.AtariGameSettings;
import com.github.tetrisanalyzer.settings.GameSettings;
import com.github.tetrisanalyzer.settings.StandardGameSettings;
import org.junit.Test;

import java.util.LinkedHashSet;
import java.util.Set;

import static com.github.tetrisanalyzer.board.Board.createBoard;
import static com.github.tetrisanalyzer.piece.Piece.createPieceI;
import static com.github.tetrisanalyzer.piece.Piece.createPieceS;
import static org.junit.Assert.assertEquals;

public class ValidPieceMovesForEmptyBoardTest {

    @Test
    public void startPieceIUsingStandardSettings() {
        Board board = createBoard(10, 20);
        GameSettings settings = new StandardGameSettings(board);
        Piece piece = createPieceI(settings);
        ValidPieceMovesForEmptyBoard validPieceMovesForEmptyBoard = new ValidPieceMovesForEmptyBoard(board, piece, settings);
        PieceMove startMove = validPieceMovesForEmptyBoard.getStartMove();

        assertEquals(new Move(0, 3, 0), startMove.move);
    }

    @Test
    public void startPieceSUsingStandardSettings() {
        Board board = createBoard(10, 20);
        GameSettings settings = new StandardGameSettings(board);
        Piece piece = createPieceS(settings);
        ValidPieceMovesForEmptyBoard validPieceMovesForEmptyBoard = new ValidPieceMovesForEmptyBoard(board, piece, settings);
        PieceMove startMove = validPieceMovesForEmptyBoard.getStartMove();

        assertEquals(new Move(0,4, 0), startMove.move);
    }

    @Test
    public void startPieceIUsingAtariSettings() {
        Board board = createBoard(10, 20);
        GameSettings settings = new AtariGameSettings(board);
        Piece piece = createPieceI(settings);
        ValidPieceMovesForEmptyBoard validPieceMovesForEmptyBoard = new ValidPieceMovesForEmptyBoard(board, piece, settings);
        PieceMove startMove = validPieceMovesForEmptyBoard.getStartMove();

        assertEquals(new Move(0,4, 1), startMove.move);
    }

    @Test
    public void startPieceSUsingAtariSettings() {
        Board board = createBoard(10, 20);
        GameSettings settings = new AtariGameSettings(board);
        Piece piece = createPieceS(settings);
        ValidPieceMovesForEmptyBoard validPieceMovesForEmptyBoard = new ValidPieceMovesForEmptyBoard(board, piece, settings);
        PieceMove startMove = validPieceMovesForEmptyBoard.getStartMove();

        assertEquals(new Move(0,4, 0), startMove.move);
    }

    @Test
    public void validMovesUsingAtariSettingsSlidingOff() {
        Board board = createBoard(5, 5);
        GameSettings settings = new AtariGameSettings(board);
        Piece piece = createPieceS(settings);
        ValidPieceMovesForEmptyBoard validPieceMovesForEmptyBoard = new ValidPieceMovesForEmptyBoard(board, piece, settings);

        Set<PieceMove> pieceMoves = new LinkedHashSet<>();
        addPieceMoves(validPieceMovesForEmptyBoard.getStartMove(), pieceMoves);

        Set<PieceMove> expectedMoves = new LinkedHashSet<>();
        expectedMoves.add(new PieceMove(board, piece, new Move(0,1, 0)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,0, 0)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,0, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,0, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,0, 3)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,1, 0)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,0, 0)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,0, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,0, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,2, 0)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,3, 0)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,3, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,3, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,2, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,2, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,1, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,1, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,2, 0)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,2, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,2, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,2, 3)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,1, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,1, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,1, 3)));

        assertEquals(expectedMoves, pieceMoves);
    }

    @Test
    public void validMovesStandardTetris() {
        Board board = createBoard(5, 5);
        GameSettings settings = new StandardGameSettings(board);
        Piece piece = createPieceS(settings);
        ValidPieceMovesForEmptyBoard validPieceMovesForEmptyBoard = new ValidPieceMovesForEmptyBoard(board, piece, settings);

        Set<PieceMove> pieceMoves = new LinkedHashSet<>();
        addPieceMoves(validPieceMovesForEmptyBoard.getStartMove(), pieceMoves);

        Set<PieceMove> expectedMoves = new LinkedHashSet<>();
        expectedMoves.add(new PieceMove(board, piece, new Move(0,1, 0)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,2, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,3, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,3, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,2, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,1, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,0, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,1, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,2, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,2, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,1, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,0, 2)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,0, 3)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,1, 3)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,2, 3)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,0, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,1, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(1,0, 1)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,0, 0)));
        expectedMoves.add(new PieceMove(board, piece, new Move(0,2, 0)));

        assertEquals(expectedMoves, pieceMoves);
    }

    private void addPieceMoves(PieceMove pieceMove, Set<PieceMove> pieceMoves) {
        if (pieceMove != null && !pieceMoves.contains(pieceMove)) {
            pieceMoves.add(pieceMove);
            for (PieceMove move : pieceMove.asideAndRotateMoves()) {
                addPieceMoves(move, pieceMoves);
            }
            addPieceMoves(pieceMove.down, pieceMoves);
        }
    }
}
